test/EngineTest.cpp

```diff
@@ -188,127 +354,112 @@ TEST(EngineTest, optionalJoinTest) {
   resWidth = vres.cols();
   CALL_FIXED_SIZE_3(aWidth, bWidth, resWidth, OptionalJoin::optionalJoin, va,
                     vb, true, false, jcls, &vres);
-
-  ASSERT_EQ(5u, vres.size());
-  ASSERT_EQ(7u, vres.cols());
-
-  vector<Id> r{I(1), I(2), I(3), I(4), I(5), I(6), I(4)};
-  for (size_t i = 0; i < 7; i++) {
-    ASSERT_EQ(r[i], vres[0][i]);
-  }
-  r = {I(1), I(2), I(3), I(4), I(5), I(6), I(5)};
-  for (size_t i = 0; i < 7; i++) {
-    ASSERT_EQ(r[i], vres[1][i]);
-  }
-  r = {I(1), I(2), I(3), I(7), I(5), I(6), I(4)};
-  for (size_t i = 0; i < 7; i++) {
-    ASSERT_EQ(r[i], vres(2, i));
-  }
-  r = {I(1), I(2), I(3), I(7), I(5), I(6), I(5)};
-  for (size_t i = 0; i < 7; i++) {
-    ASSERT_EQ(r[i], vres(3, i));
-  }
-  r = {ID_NO_VALUE, I(6), I(7), ID_NO_VALUE, ID_NO_VALUE, ID_NO_VALUE, I(4)};
-  for (size_t i = 0; i < 7; i++) {
-    ASSERT_EQ(r[i], vres(4, i));
-  }
+  
+  // For easier checking.
+  sampleSolution = makeIdTableFromVector(std::vector<std::array<size_t, 7>>{
+          {1, 2, 3, 4, 5, 6, 4},
+          {1, 2, 3, 4, 5, 6, 5},
+          {1, 2, 3, 7, 5, 6, 4},
+          {1, 2, 3, 7, 5, 6, 5},
+          {0, 6, 7, 0, 0, 0, 4}
+        }
+      );
+  sampleSolution(4, 0) = ID_NO_VALUE;
+  sampleSolution(4, 3) = ID_NO_VALUE;
+  sampleSolution(4, 4) = ID_NO_VALUE;
+  sampleSolution(4, 5) = ID_NO_VALUE;
+
+  ASSERT_EQ(sampleSolution.size(), vres.size());
+  ASSERT_EQ(sampleSolution.cols(), vres.cols());
+  ASSERT_EQ(sampleSolution, vres);
```

Comment:
I think we can even generalize your mechanism from above, s.t. you can also use your functions `goThroughTestCases` for the `optionalJoin` and the `minus` (if that is even in here). 
The difficulty there is, that we need to generalize the notion of the `JoinColumns`, but we can make this work.
Maybe first perform the other suggestions.

Reply:
It would be doable, but I'm not sure if the extra work needed for re-working the structure for test cases and re-designing the help functions for `goThroughTestCases` would be worth it.

After all, the cases of `optionalJoin` don't get re-used anywhere else and aren't very big in the first place.